@supports(background-attachment:fixed) should return false on iOS

zakariaridouh opened the pull request:
#### 240b0227399a2330bc1704680c2dd2f1769a96b3
<pre>
@supports(background-attachment:fixed) should return false on iOS
<a href="https://bugs.webkit.org/show_bug.cgi?id=196327">https://bugs.webkit.org/show_bug.cgi?id=196327</a>
<a href="https://rdar.apple.com/95035888">rdar://95035888</a>

Reviewed by NOBODY (OOPS!).

Properly gate support for background-attachment:fixed property on iOS.

Rename the settings flag for support from fixedBackgroundsPaintRelativeToDocument to cssFixedBackgroundSupportEnabled and flip flag for iOS.
This gives a clearer idea to what the flag is gating, and properly returns false when @supports(background-attachment:fixed) is resolved on iOS.

Write custom CSS parser to properly plumb the support across the codebase.

Added tests to verify this functionality.

  * LayoutTests/fast/css/fixed-background-support-disabled-expected.txt: Added.
  * LayoutTests/fast/css/fixed-background-support-disabled.html: Added.
  * LayoutTests/fast/css/fixed-background-support-enabled-expected.txt: Added.
  * LayoutTests/fast/css/fixed-background-support-enabled.html: Added.
  * Source/WebCore/css/CSSProperties.json:
  * Source/WebCore/css/parser/CSSParserContext.cpp:
  (WebCore::add):
  * Source/WebCore/css/parser/CSSParserContext.h:
  * Source/WebCore/css/parser/CSSPropertyParser.cpp:
  (WebCore::consumeBackgroundComponent):
  * Source/WebCore/css/parser/CSSPropertyParserHelpers.cpp:
  (WebCore::CSSPropertyParserHelpers::consumeSingleBackgroundAttachment):
  * Source/WebCore/css/parser/CSSPropertyParserHelpers.h:
  * Source/WebCore/page/Settings.yaml:
  * Source/WebCore/rendering/BackgroundPainter.cpp:
  (WebCore::BackgroundPainter::calculateBackgroundImageGeometry):
  * Source/WebCore/rendering/RenderElement.cpp:
  (WebCore::RenderElement::styleWillChange):
  (WebCore::RenderElement::willBeDestroyed):
  * Source/WebCore/rendering/RenderLayerCompositor.cpp:
  (WebCore::RenderLayerCompositor::needsFixedRootBackgroundLayer const):
  (WebCore::RenderLayerCompositor::updateSynchronousScrollingNodes):
</pre><!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/240b0227399a2330bc1704680c2dd2f1769a96b3

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/63811 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/43168 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/55/builds/16408 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/67832 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/59/builds/14419 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/65931 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/50858 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/61/builds/14699 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/51418 "Passed tests") | [✅ 🧪 wincairo-tests](https://ews-build.webkit.org/#/builders/60/builds/9970 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/66880 "Passed tests") | [❌ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/39985 "Found 2 new test failures: fast/backgrounds/multiple-backgrounds-computed-style.html fast/css/getComputedStyle/getComputedStyle-background-shorthand.html (failure)") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/55243 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/32035 "Passed tests") | 
| | [❌ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/42/builds/36671 "Found 15 new test failures: imported/w3c/web-platform-tests/css/css-backgrounds/animations/discrete-no-interpolation.html imported/w3c/web-platform-tests/css/css-backgrounds/background-origin-006.html imported/w3c/web-platform-tests/css/css-backgrounds/inheritance.sub.html imported/w3c/web-platform-tests/css/css-backgrounds/parsing/background-attachment-computed.html imported/w3c/web-platform-tests/css/css-backgrounds/parsing/background-attachment-valid.html imported/w3c/web-platform-tests/css/css-backgrounds/parsing/background-computed.html imported/w3c/web-platform-tests/css/css-cascade/all-prop-revert-layer.html imported/w3c/web-platform-tests/css/css-pseudo/first-letter-allowed-properties.html imported/w3c/web-platform-tests/css/css-pseudo/first-line-allowed-properties.html imported/w3c/web-platform-tests/css/css-typed-om/the-stylepropertymap/properties/background-attachment.html ... (failure)") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/64/builds/12624 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/13292 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/58641 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/63/builds/12951 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/69528 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/87/builds/7758 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/62/builds/12494 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/58741 "Passed tests") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/86/builds/7791 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/55339 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/58885 "Passed tests") | 
| | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/88/builds/6451 "Passed tests") | | | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/44/builds/38988 "Built successfully") | | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/46/builds/40067 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/43/builds/41250 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/45/builds/39810 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

smfr commented on Source/WebCore/page/Settings.yaml:
```diff
@@ -115,8 +115,8 @@ FixedBackgroundsPaintRelativeToDocument:
     WebCore:
-      PLATFORM(IOS_FAMILY): true
-      default: false
+      PLATFORM(IOS_FAMILY): false
```
Why did you flip this around? "fixedBackgroundsPaintRelativeToDocument" is the iOS behavior

zakariaridouh replied:
Switched to "cssFixedBackgroundSupport" for clarity.

nt1m commented on Source/WebCore/css/CSSProperties.json:
```diff
@@ -1586,8 +1586,11 @@
+                "local",
+                {
+                    "value": "fixed",
+                    "settings-flag": "fixedBackgroundsPaintRelativeToDocument"
```
cssFixedBackgroundSupport

nt1m commented on Source/WebCore/page/Settings.yaml:
```diff
@@ -115,8 +115,8 @@ FixedBackgroundsPaintRelativeToDocument:
   type: bool
```
CSSFixedBackgroundSupport

nt1m commented on LayoutTests/fast/css/fixed-background-support-disabled-expected.txt:
```diff
@@ -0,0 +1,15 @@
+
+PASS 'backgroundAttachment' in document.documentElement.style is true
+PASS 'background-attachment' in getComputedStyle(document.documentElement) is true
+FAIL CSS.supports('background-attachment:fixed') should be false. Was true.
```
This doesn't look right

nt1m commented on Source/WebCore/css/parser/CSSPropertyParserHelpers.cpp:
```diff
@@ -3222,6 +3222,27 @@ RefPtr<CSSValue> consumeRepeatStyle(CSSParserTokenRange& range, const CSSParserC
+        range.consumeIncludingWhitespace();
+        return CSSPrimitiveValue::create(keyword);
+    // Gate "background-attachment: fixed" at a settings flag level.
+    case CSSValueID::CSSValueFixed:
```
nit: drop the optional `CSSValueID::` prefix

nt1m commented on Source/WebCore/css/process-css-properties.py:
Can we drop all the changes in this file?

nt1m commented on LayoutTests/fast/css/fixed-background-support-disabled.html:
```diff
@@ -0,0 +1,27 @@
+shouldBeTrue("'backgroundAttachment' in document.documentElement.style");
+shouldBeTrue("'background-attachment' in getComputedStyle(document.documentElement)");
+
+// make sure @supports(background-attachment:fixed) returns NO when CSSFixedBackgroundSupportEnabled is disabled.
```
```suggestion
// Make sure @supports(background-attachment:fixed) returns false when CSSFixedBackgroundSupportEnabled is disabled.
```

nt1m commented on LayoutTests/fast/css/fixed-background-support-disabled.html:
```diff
@@ -0,0 +1,27 @@
+// make sure @supports(background-attachment:fixed) returns NO when CSSFixedBackgroundSupportEnabled is disabled.
+shouldBeFalse("CSS.supports('background-attachment:fixed')");
+
+// make sure the flag does not impact the other kinds of background-attachment.
```
```suggestion
// Make sure the flag does not impact the other kinds of background-attachment.
```

nt1m commented on LayoutTests/fast/css/fixed-background-support-disabled.html:
```diff
@@ -0,0 +1,27 @@
+// make sure the flag does not impact the other kinds of background-attachment.
+shouldBeTrue("CSS.supports('background-attachment:scroll')");
+shouldBeTrue("CSS.supports('background-attachment:local')");
+
```
This blank line can be removed

nt1m commented on LayoutTests/fast/css/fixed-background-support-enabled.html:
```diff
@@ -0,0 +1,24 @@
+shouldBeTrue("CSS.supports('background-attachment:fixed')");
+shouldBeTrue("CSS.supports('background-attachment:scroll')");
+shouldBeTrue("CSS.supports('background-attachment:local')");
+
```
This blank line can be removed.

nt1m commented on LayoutTests/fast/css/fixed-background-support-disabled.html:
```diff
@@ -0,0 +1,27 @@
+
+</script>
+</body>
+</html>
```
Can you add the missing blank line at the end of the file?

nt1m commented on LayoutTests/fast/css/fixed-background-support-enabled.html:
```diff
@@ -0,0 +1,24 @@
+
+</script>
+</body>
+</html>
```
ditto

nt1m commented on Source/WebCore/css/parser/CSSPropertyParserHelpers.cpp:
```diff
@@ -3222,6 +3222,27 @@ RefPtr<CSSValue> consumeRepeatStyle(CSSParserTokenRange& range, const CSSParserC
+            range.consumeIncludingWhitespace();
+            return CSSPrimitiveValue::create(keyword);
+        }
+        break;
```
the duplication can be avoided using the suggestion from Slack

nt1m commented on Source/WebCore/css/parser/CSSPropertyParser.cpp:
```diff
@@ -1809,6 +1809,7 @@ static bool consumeBackgroundPosition(CSSParserTokenRange& range, const CSSParse
     return true;
 }
 
+enum class FromShorthand : bool { No, Yes };
```
I'd move this in `Source/WebCore/css/parser/CSSPropertyParserHelpers.h`

nt1m commented on Source/WebCore/css/parser/CSSPropertyParserHelpers.cpp:
```diff
@@ -3222,6 +3222,23 @@ RefPtr<CSSValue> consumeRepeatStyle(CSSParserTokenRange& range, const CSSParserC
     return CSSBackgroundRepeatValue::create(*value1, *value2);
 }
 
+RefPtr<CSSValue> consumeSingleBackgroundAttachment(CSSParserTokenRange& range, const CSSParserContext& context, const FromShorthand fromShorthand = FromShorthand::No)
```
nit: drop `const` from shorthand parameter

nt1m commented on Source/WebCore/css/parser/CSSPropertyParserHelpers.h:
```diff
@@ -184,6 +184,7 @@ RefPtr<CSSValue> consumeBorderImageWidth(CSSPropertyID, CSSParserTokenRange&);
 bool consumeBorderImageComponents(CSSPropertyID, CSSParserTokenRange&, const CSSParserContext&, RefPtr<CSSValue>&, RefPtr<CSSValue>&, RefPtr<CSSValue>&, RefPtr<CSSValue>&, RefPtr<CSSValue>&);
 RefPtr<CSSValue> consumeWebkitBorderImage(CSSPropertyID, CSSParserTokenRange&, const CSSParserContext&);
 RefPtr<CSSValue> consumeReflect(CSSParserTokenRange&, const CSSParserContext&);
+RefPtr<CSSValue> consumeSingleBackgroundAttachment(CSSParserTokenRange&, const CSSParserContext&, const FromShorthand);
```
ditto: drop const here

zakariaridouh committed:
@supports(background-attachment:fixed) should return false on iOS
https://bugs.webkit.org/show_bug.cgi?id=196327
rdar://95035888

Reviewed by NOBODY (OOPS!).

Properly gate support for background-attachment:fixed property on iOS.

Rename the settings flag for support from fixedBackgroundsPaintRelativeToDocument to cssFixedBackgroundSupportEnabled and flip flag for iOS.
This gives a clearer idea to what the flag is gating, and properly returns false when @supports(background-attachment:fixed) is resolved on iOS.

Write custom CSS parser to properly plumb the support across the codebase.

Added tests to verify this functionality.

  * LayoutTests/fast/css/fixed-background-support-disabled-expected.txt: Added.
  * LayoutTests/fast/css/fixed-background-support-disabled.html: Added.
  * LayoutTests/fast/css/fixed-background-support-enabled-expected.txt: Added.
  * LayoutTests/fast/css/fixed-background-support-enabled.html: Added.
  * Source/WebCore/css/CSSProperties.json:
  * Source/WebCore/css/parser/CSSParserContext.cpp:
  (WebCore::add):
  * Source/WebCore/css/parser/CSSParserContext.h:
  * Source/WebCore/css/parser/CSSPropertyParser.cpp:
  (WebCore::consumeBackgroundComponent):
  * Source/WebCore/css/parser/CSSPropertyParserHelpers.cpp:
  (WebCore::CSSPropertyParserHelpers::consumeSingleBackgroundAttachment):
  * Source/WebCore/css/parser/CSSPropertyParserHelpers.h:
  * Source/WebCore/page/Settings.yaml:
  * Source/WebCore/rendering/BackgroundPainter.cpp:
  (WebCore::BackgroundPainter::calculateBackgroundImageGeometry):
  * Source/WebCore/rendering/RenderElement.cpp:
  (WebCore::RenderElement::styleWillChange):
  (WebCore::RenderElement::willBeDestroyed):
  * Source/WebCore/rendering/RenderLayerCompositor.cpp:
  (WebCore::RenderLayerCompositor::needsFixedRootBackgroundLayer const):
  (WebCore::RenderLayerCompositor::updateSynchronousScrollingNodes):
